### PR TITLE
✏️ Edit profile accounts

### LINKED ISSUE #207: Edit my profile details
@[figma](https://www.figma.com/file/J2KUqGu97lNQC2iKsxJeVH/My-Profile?node-id=192%3A14888)

### Requirements

- Display inline form for modyfiyng member details: handle, name, about
- Display sign&authorize modal for given changes (should have transaction fee displayed).

### LINKED ISSUE #208: Submit profile update transaction & handle response

### PR DESCRIPTION
**Pull request by jodator**
Closes #207
Closes #208

### CODE REVIEW

File: packages/ui/src/modals/UpdateMembershipModal/types.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  controllerAccount?: Account
+}
+
+export type Action<T> = {
+  type: keyof T
+  value?: T[keyof T]
+}
+export type FormReducer<T> = Reducer<T, Action<T>>
+
+export type Nullable<T> = { [P in keyof T]: T[P] | null }
```
Comment (krzysztof-jelski) [inline packages/ui/src/modals/UpdateMembershipModal/types.ts:20]: I don't like the `Nullable` name, as the name suggests to me the type is something like:
```type Nullable<T> = T | null```
But frankly I don't have a good suggestion. `WithNullableValues`?

File: packages/ui/src/modals/UpdateMembershipModal/SignUpdateMembershipModal.tsx [hunk 2]
```diff
@@ -11,33 +13,56 @@ import { useApi } from '../../hooks/useApi'
...
+  return fields.some((field) => !!object[field])
+}
+
+function createBatch(transactionParams: Nullable<UpdateMemberForm>, api: ApiRx | undefined, member: BaseMember) {
+  const hasProfileEdits = hasEdits(transactionParams, ['about', 'handle', 'avatarURI', 'name'])
+  const hasAccountsEdits = hasEdits(transactionParams, ['rootAccount', 'controllerAccount'])
+
+  const transactions: SubmittableExtrinsic<'rxjs'>[] = []
+
+  if (!api || (!hasProfileEdits && !hasAccountsEdits)) {
```
Comment (krzysztof-jelski) [inline packages/ui/src/modals/UpdateMembershipModal/SignUpdateMembershipModal.tsx:36]:
```suggestion
  if (!api || !(hasProfileEdits || hasAccountsEdits)) {
```
Not that it's ultimately better, but for me personally this reads easier.

File: packages/ui/src/modals/UpdateMembershipModal/SignUpdateMembershipModal.tsx [hunk 3]
```diff
@@ -11,33 +13,56 @@ import { useApi } from '../../hooks/useApi'
...
+    return
+  }
+  if (hasProfileEdits) {
+    const updateProfile = api.tx.members.updateProfile(
+      member.id,
+      transactionParams.name || null,
```
Comment (krzysztof-jelski) [inline packages/ui/src/modals/UpdateMembershipModal/SignUpdateMembershipModal.tsx:42]: The API would handle `undefined` fine, no?

File: packages/ui/src/modals/UpdateMembershipModal/utils.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+export const hasAnyEdits = (formData: Record<string, any>, member: Record<string, any>) => {
+  return !!getChangedFields(formData, member).length
+}
+
+export const getChangedFields = (form: Record<string, any>, initial: Record<string, any>) => {
+  const changedFields = []
+
+  for (const key of Object.keys(form)) {
+    const memberValue = initial[key] || ''
```
Comment (krzysztof-jelski) [inline packages/ui/src/modals/UpdateMembershipModal/utils.ts:11]:
```suggestion
    const initialValue = initial[key] || ''
```

**Review by krzysztof-jelski — APPROVED** [review]

File: packages/ui/src/modals/UpdateMembershipModal/SignUpdateMembershipModal.tsx (hunk 3 above)
Comment (jodator) [reply packages/ui/src/modals/UpdateMembershipModal/SignUpdateMembershipModal.tsx:42]: I'll check that, but AFAIR there was `null` not undefined.

Comment (jodator) [reply packages/ui/src/modals/UpdateMembershipModal/SignUpdateMembershipModal.tsx:42]: Yep, in the handbook, all the `updateProfile()` parameters are described as optional and the type for them is `Option<Bytes> | null | object | string | Uint8Array`.

File: packages/ui/src/modals/UpdateMembershipModal/types.ts (hunk 1 above)
Comment (jodator) [reply packages/ui/src/modals/UpdateMembershipModal/types.ts:20]: Sounds better, I'll use that :+1:

**Pull request merged**